From a68689d77ce5df1378040492d9e982774bbbade1 Mon Sep 17 00:00:00 2001
From: Jerry Wang <Jerry.Wang4@arm.com>
Date: Thu, 19 Dec 2024 20:07:59 +0000
Subject: [PATCH] fix(rdv3r1): add console name to checksum calculation on
 RD-V3-R1

The name field of console_info structure was missed in
checksum calculation. This is corrected by adding a new
helper checksum_calc() which computes the checksum in a
field agnostic manner.

Signed-off-by: Jerry Wang <Jerry.Wang4@arm.com>
Change-Id: I318379f3a64f26b7f84c6d162d2d5a3085f5c9a0

Signed-off-by: Jon Mason <jon.mason@arm.com>
Upstream-Status: Pending
---
 .../platform/rdv3r1/rdv3r1_common.c           | 33 +++++++++++++++----
 1 file changed, 26 insertions(+), 7 deletions(-)

diff --git a/plat/arm/board/neoverse_rd/platform/rdv3r1/rdv3r1_common.c b/plat/arm/board/neoverse_rd/platform/rdv3r1/rdv3r1_common.c
index 9c952d97cab6..cb63a25e7de0 100644
--- a/plat/arm/board/neoverse_rd/platform/rdv3r1/rdv3r1_common.c
+++ b/plat/arm/board/neoverse_rd/platform/rdv3r1/rdv3r1_common.c
@@ -32,6 +32,23 @@ unsigned int plat_arm_nrd_get_multi_chip_mode(void)
 }
 
 #if ENABLE_RME
+/*
+ * Calculate checksum of 64-bit words @buffer with @size length
+ */
+static uint64_t checksum_calc(uint64_t *buffer, size_t size)
+{
+	uint64_t sum = 0UL;
+
+	assert(((uintptr_t)buffer & (sizeof(uint64_t) - 1UL)) == 0UL);
+	assert((size & (sizeof(uint64_t) - 1UL)) == 0UL);
+
+	for (unsigned long i = 0UL; i < (size / sizeof(uint64_t)); i++) {
+		sum += buffer[i];
+	}
+
+	return sum;
+}
+
 /*
  * Get a pointer to the RMM-EL3 shared buffer and return it
  * through the pointer passed as parameter.
@@ -138,8 +155,8 @@ int plat_rmmd_load_manifest(struct rmm_manifest *manifest)
 	bank_ptr[1].size = ARM_DRAM2_SIZE;
 
 	/* Update checksum */
-	checksum += bank_ptr[0].base + bank_ptr[0].size + bank_ptr[1].base +
-		bank_ptr[1].size;
+	checksum += checksum_calc((uint64_t *)bank_ptr,
+		sizeof(struct ns_dram_bank) * num_banks);
 
 	/* Checksum must be 0 */
 	manifest->plat_dram.checksum = ~checksum + 1UL;
@@ -148,18 +165,20 @@ int plat_rmmd_load_manifest(struct rmm_manifest *manifest)
 	checksum = num_consoles + (uint64_t)console_ptr;
 
 	/* Zero out the console info struct */
-	memset((void *)console_ptr, '\0', sizeof(struct console_info) * num_consoles);
+	(void)memset((void *)console_ptr, '\0',
+		sizeof(struct console_info) * num_consoles);
 
-	console_ptr[0].map_pages = 1;
+	console_ptr[0].map_pages = 1UL;
 	console_ptr[0].base = NRD_CSS_RMM_CONSOLE_BASE;
 	console_ptr[0].clk_in_hz = NRD_CSS_RMM_CONSOLE_CLK_IN_HZ;
 	console_ptr[0].baud_rate = NRD_CSS_RMM_CONSOLE_BAUD;
 
-	strlcpy(console_ptr[0].name, NRD_CSS_RMM_CONSOLE_NAME, RMM_CONSOLE_MAX_NAME_LEN-1UL);
+	(void)strlcpy(console_ptr[0].name, NRD_CSS_RMM_CONSOLE_NAME,
+		RMM_CONSOLE_MAX_NAME_LEN-1UL);
 
 	/* Update checksum */
-	checksum += console_ptr[0].base + console_ptr[0].map_pages +
-		console_ptr[0].clk_in_hz + console_ptr[0].baud_rate;
+	checksum += checksum_calc((uint64_t *)console_ptr,
+		sizeof(struct console_info) * num_consoles);
 
 	/* Checksum must be 0 */
 	manifest->plat_console.checksum = ~checksum + 1UL;
